cpp/src/rolling/range_window_bounds_detail.hpp

```diff
@@ -0,0 +1,212 @@
+  {
+    using order_by_column_duration_t = typename OrderByColumnType::duration;
+    using rep_t                      = typename order_by_column_duration_t::rep;
+
+    auto const& range_scalar_duration =
+      static_cast<cudf::duration_scalar<RangeType> const&>(range_scalar);
+    return std::unique_ptr<scalar>{new cudf::duration_scalar<order_by_column_duration_t>{
+      is_unbounded_range ? order_by_column_duration_t{std::numeric_limits<rep_t>::max()}
+                         : order_by_column_duration_t{range_scalar_duration.value(stream)},
+      true,
+      stream,
+      mr}};
+  }
+
+  template <typename RangeType,
+            std::enable_if_t<std::is_same<OrderByColumnType, RangeType>::value &&
+                               std::is_integral<RangeType>::value && !cudf::is_boolean<RangeType>(),
+                             void>* = nullptr>
+  std::unique_ptr<scalar> operator()(scalar const& range_scalar,
+                                     bool is_unbounded_range,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr) const
+  {
+    using numeric_scalar = cudf::numeric_scalar<RangeType>;
+
+    return std::unique_ptr<scalar>{new numeric_scalar{
+      is_unbounded_range ? std::numeric_limits<RangeType>::max()
+                         : static_cast<numeric_scalar const&>(range_scalar).value(stream),
+      true,
+      stream,
+      mr}};
+  }
+};
+
+namespace {
+template <typename RepType>
+struct range_comparable_value_fetcher {
+  template <typename RangeType, typename... Args>
+  std::enable_if_t<!is_supported_range_type<RangeType>(), RepType> operator()(Args&&...) const
+  {
+    CUDF_FAIL("Unsupported window range type!");
+  }
+
+  template <typename RangeType>
+  std::enable_if_t<std::is_integral<RangeType>::value && !cudf::is_boolean<RangeType>(), RepType>
+  operator()(scalar const& range_scalar, rmm::cuda_stream_view stream) const
+  {
+    return static_cast<numeric_scalar<RangeType> const&>(range_scalar).value(stream);
+  }
+
+  template <typename RangeType>
+  std::enable_if_t<cudf::is_duration<RangeType>(), RepType> operator()(
+    scalar const& range_scalar, rmm::cuda_stream_view stream) const
+  {
+    return static_cast<duration_scalar<RangeType> const&>(range_scalar).value(stream).count();
+  }
+};
+
+template <typename RepType>
+bool rep_type_compatible_for_range_comparison(type_id id)
+{
+  return (id == type_id::DURATION_DAYS && std::is_same<RepType, int32_t>()) ||
+         (id == type_id::DURATION_SECONDS && std::is_same<RepType, int64_t>()) ||
+         (id == type_id::DURATION_MILLISECONDS && std::is_same<RepType, int64_t>()) ||
+         (id == type_id::DURATION_MICROSECONDS && std::is_same<RepType, int64_t>()) ||
+         (id == type_id::DURATION_NANOSECONDS && std::is_same<RepType, int64_t>()) ||
+         type_id_matches_device_storage_type<RepType>(id);
+};
```

Comment:
Is there any way to avoid such manually adding types? Such as combining constexpr check like `is_duration<>()` and `device_storage_type`?

Reply:
That would destabilize what `device_storage_type` is currently checking for, i.e. decimals. This is specific to the context of range types.

Reply:
Hmm... I thought I'd replied (incorrectly) here.

I see what your suggestion is. Let me give that a go.

Reply:
Nope. This doesn't seem possible without yet another type dispatch. `id` is a runtime value.

Even if we used `is_duration(data_type)` for the first part of each clause, we'd need a dispatch to fetch the duration's rep-type for the second part of the clause.

Looks like this was the reason why `type_id_matches_device_storage_type<>()` also has an explicit list. I would have liked `type_id_matches_device_storage_type()` to have included durations/timestamps, I think that would have been incorrect: chrono types store chrono values, not integrals.